### PR TITLE
Add some DataBox protocols

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

work towards #4761 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by kidder** [comment]
@sxs-collaboration/spectre-core-devs I would appreciate some feedback before I proceed further...

File: src/DataStructures/DataBox/Protocols/Mutator.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+///      - A `const db::const_item_type<Tag, BoxTags>` for each `Tag` in
+///        `argument_tags`
+///      - The additional arguments passed to `db::mutate_apply`
+struct Mutator {
+  template <typename ConformingType>
+  struct test {
+    using argument_tags = typename ConformingType::argument_tags;
+    using return_tags = typename ConformingType::return_tags;
+    static_assert(
+        std::is_same_v<void, std::void_t<decltype(&ConformingType::apply)>>);
```
Comment (nilsvu) [inline src/DataStructures/DataBox/Protocols/Mutator.hpp:34]: The apply function might be a template, in which case I think this won't work. Maybe do a quick test and delete this assert if it doesn't work? Also I think `db::mutate_apply` also works with `operator()`. Please add that to the docs.

File: src/DataStructures/DataBox/Protocols/Mutator.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/// - `return_tags`: A type list of tags corresponding to mutable items in the
+///   DataBox passed to `db::mutate_apply` that may be modified.
+/// - `argument_tags`: A type list of tags corresponding to items in the DataBox
+///   passed to `db::mutate_apply` that may not be modified.
+/// - `apply`: A static function whose return value is returned by
+///   `db::mutate_apply`, and that takes as arguments:
+///      - A `const gsl::not_null<Tag::type*>` for each `Tag` in `return_tags`
+///      - A `const db::const_item_type<Tag, BoxTags>` for each `Tag` in
+///        `argument_tags`
+///      - The additional arguments passed to `db::mutate_apply`
```
Comment (nilsvu) [inline src/DataStructures/DataBox/Protocols/Mutator.hpp:27]: Add an example snippet please

**Review by nilsvu — CHANGES_REQUESTED** [review]
I think the mutator protocol looks great, that's exactly what we need more of. It has a clear name and documents a compile-time interface that people would otherwise have a lot of trouble understanding.

Example: Just yesterday Mike and I tried to figure out what the interface for "variable fixers" is. It's not documented (at least we didn't find any docs). Only when we read the source code of the FixVariables action did we realize that a `VariableFixer` just needs to be (1) a mutator and (2) option-creatable. It would be so much clearer if those two were protocols.

My suggestion: Can we limit this PR to the `Mutator` protocol and merge that first because it's very clear, then move on to the others?

**Pull request closed without merging**